### PR TITLE
app_voicemail: add CLI commands for message manipulation

### PR DESCRIPTION
**Pull request by mbradeen**
Adds CLI commands to allow move/remove/forward individual messages
from a particular mailbox folder. The forward command can be used
to copy a message within a mailbox or to another mailbox. Also adds
a show mailbox, required to retrieve message ID's.

Resolves: #170

UserNote: The following CLI commands have been added to app_voicemail

voicemail show mailbox \<mailbox> \<context>
Show contents of mailbox \<mailbox>\@\<context>

voicemail remove \<mailbox> \<context> \<from_folder> \<messageid>
Remove message \<messageid> from \<from_folder> in mailbox \<mailbox> \@ \<context>

voicemail move \<mailbox> \<context> \<from_folder> \<messageid> \<to_folder>
Move message \<messageid> in mailbox \<mailbox> \@ \<context> from \<from_folder> to \<to_folder>

voicemail forward \<from_mailbox> \<from_context> \<from_folder> \<messageid> \<to_mailbox> \<to_context> \<to_folder>
Forward message \<messageid> in mailbox \<mailbox> \@ \<context> \<from_folder> to
mailbox \<mailbox> \@ \<context> \<to_folder>

### CODE REVIEW

**Comment by mbradeen** [comment]
cherry-pick-to: 20
cherry-pick-to: 18
cherry-pick-to: certified/18.9

File: apps/app_voicemail.c [hunk 1]
```diff
@@ -11413,6 +11412,371 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+	const char *mailbox = a->argv[3];
+	const char *context = a->argv[4];
+	struct ast_vm_mailbox_snapshot *mailbox_snapshot;
+	struct ast_vm_msg_snapshot *msg;
+
+	/* Take a snapshot of the mailbox and walk through each folder's contents */
+	if ((mailbox_snapshot = ast_vm_mailbox_snapshot_create(mailbox, context, NULL, 0, AST_VM_SNAPSHOT_SORT_BY_ID, 0))) {
+		ast_cli(a->fd, VM_STRING_HEADER_FORMAT, "Folder", "Caller ID", "Date", "Duration", "Flag", "ID");
+
+		for (int i=0; i<mailbox_snapshot->folders; i++) {
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11451]: Add spaces between operators, e.g. = and <

File: apps/app_voicemail.c [hunk 2]
```diff
@@ -11413,6 +11412,371 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+					ret = ast_strdup(vmu->mailbox);
+					AST_LIST_UNLOCK(&users);
+					return ret;
+				}
+				mailbox = vmu->mailbox;
+			}
+		}
+		AST_LIST_UNLOCK(&users);
+	}
+	else if (pos == 4) {
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11535]: move else if to previous line

File: apps/app_voicemail.c [hunk 3]
```diff
@@ -11413,6 +11412,371 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+		/* only display contexts that match the user in pos 3*/
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11536]: space before end of comment

File: apps/app_voicemail.c [hunk 4]
```diff
@@ -11413,6 +11412,371 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+					ret = ast_strdup(vmu->mailbox);
+					AST_LIST_UNLOCK(&users);
+					return ret;
+				}
+				mailbox = vmu->mailbox;
+			}
+		}
+		AST_LIST_UNLOCK(&users);
+	}
+	else if (pos == 3 || pos == 7) {
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11629]: previous line

File: apps/app_voicemail.c [hunk 5]
```diff
@@ -11413,6 +11412,371 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+		/* find contexts that match the user */
+		if (pos == 3) {
+			mailbox = a->argv[2];
+		}
+		else {
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11634]: previous line

File: apps/app_voicemail.c [hunk 6]
```diff
@@ -11413,6 +11412,371 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+					ret = ast_strdup(vmu->context);
+					AST_LIST_UNLOCK(&users);
+					return ret;
+				}
+				context = vmu->context;
+			}
+		}
+		AST_LIST_UNLOCK(&users);
+	}
+	else if (pos == 4 || pos == 8 || (pos == 6 && maxpos == 6) ) {
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11652]: prev line

File: apps/app_voicemail.c [hunk 7]
```diff
@@ -11413,6 +11412,371 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+					ret = ast_strdup(msg->msg_id);
+					break;
+				}
+				id = msg->msg_id;
+			}
+
+			/* done, destroy. */
+			mailbox_snapshot = ast_vm_mailbox_snapshot_destroy(mailbox_snapshot);
+		}
+		else {
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11691]: prev line

File: apps/app_voicemail.c [hunk 8]
```diff
@@ -11413,6 +11412,371 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+			ast_log(LOG_NOTICE, "Can't create snapshot for voicemail box %s@%s\n", mailbox, context);
+		}
+
+		return ret;
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11695]: This return is redundant and can be removed

File: apps/app_voicemail.c [hunk 9]
```diff
@@ -12771,20 +13135,26 @@ static char *complete_voicemail_show_users(const char *line, const char *word, i
...
+				AST_LIST_UNLOCK(&users);
+				return ret;
+			}
 			/* ignore repeated contexts ? */
 			context = vmu->context;
 		}
 	}
-	return NULL;
+	AST_LIST_UNLOCK(&users);
+	return ret;
```
Comment (InterLinked1) [inline apps/app_voicemail.c:13157]: I'm not following why you can't just return NULL here. In which case the initialize of ret to NULL is also unnecessary.

File: apps/app_voicemail.c (hunk 3 above)
Comment (InterLinked1) [inline apps/app_voicemail.c:11536]: ``<sidenote>``
Personally, I feel like it would be more logical to show all mailboxes in a given context, than all contexts that could match a given mailbox... but I guess this fits with mailbox being required and context defaulting to "default" and thus being optional. Just feels a bit bizarre that it works for me.
``</sidenote>``

**Review by InterLinked1 — CHANGES_REQUESTED** [review]

File: apps/app_voicemail.c (hunk 3 above)
Comment (mbradeen) [reply apps/app_voicemail.c]: It does make sense that context is major over user, but people think of the mailbox as user@context and I think it would make the syntax confusing to put them in the other order.  Even if from a sorting perspective it's the other way around.

File: apps/app_voicemail.c [hunk 10]
```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+
+	const char *mailbox = a->argv[3];
+	const char *context = a->argv[4];
+	struct vm_state vms;
+	struct ast_vm_user *vmu = NULL, vmus;
+	memset(&vmus, 0, sizeof(vmus));
+	memset(&vms, 0, sizeof(vms));
+
+	if (!(vmu = find_user(&vmus, context, mailbox))) {
+		ast_log(LOG_NOTICE, "Can't find voicemail user %s@%s\n", mailbox, context);
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11428]: This should not be a NOTICE message. It should use ast_cli.

File: apps/app_voicemail.c [hunk 11]
```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+				ast_cli(a->fd, VM_STRING_DATA_FORMAT, msg->folder_name, msg->callerid, msg->origdate, msg->duration,
+						msg->flag, msg->msg_id);
+			}
+		}
+
+		ast_cli(a->fd, "%d Messages Total\n", mailbox_snapshot->total_msg_num);
+		/* done, destroy. */
+		mailbox_snapshot = ast_vm_mailbox_snapshot_destroy(mailbox_snapshot);
+	} else {
+		ast_log(LOG_NOTICE, "Can't create snapshot for voicemail box %s@%s\n", mailbox, context);
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11462]: Is this an actual issue, or something that could happen legitimately? If not, it should be a warning.

File: apps/app_voicemail.c [hunk 12]
```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+		return complete_voicemail_show_mailbox(a);
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 5) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (show_messages_for_mailbox(a)){
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11574]: space needed before {

File: apps/app_voicemail.c [hunk 13]
```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+			mailbox = a->argv[2];
+		} else {
+			mailbox = a->argv[6];
+		}
+
+		wordlen = strlen(word);
+		AST_LIST_LOCK(&users);
+		AST_LIST_TRAVERSE(&users, vmu, list) {
+			if (!strncasecmp(word, vmu->context, wordlen) && !strcasecmp(mailbox, vmu->mailbox)) {
+				if (context && strcmp(context, vmu->context) && ++which > state){
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11642]: ){ whitespace

File: apps/app_voicemail.c [hunk 14]
```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+		return complete_voicemail_move_message(a, 8);
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 9) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (forward_message_from_mailbox(a)){
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11715]: ){ whitespace

File: apps/app_voicemail.c [hunk 15]
```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+		return complete_voicemail_move_message(a, 6);
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 7) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (move_message_from_mailbox(a)){
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11741]: ){ whitespace

File: apps/app_voicemail.c [hunk 16]
```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+		return complete_voicemail_move_message(a, 5);
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 6) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (remove_message_from_mailbox(a)){
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11767]: ){ space

File: apps/app_voicemail.c [hunk 17]
```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 7) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (move_message_from_mailbox(a)){
+		ast_log(AST_LOG_WARNING, "Error moving message!\n");
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11742]: Probably should be an error, not a warning, and it should include details of what message move failed.

**Review by InterLinked1 — CHANGES_REQUESTED** [review]

File: apps/app_voicemail.c [hunk 18]
```diff
@@ -6184,7 +6184,6 @@ static int msg_create_from_file(struct ast_vm_recording_data *recdata)
 	#ifdef IMAP_STORAGE
 	struct vm_state *vms = NULL;
 	char ext_context[256] = "";
-	char *fmt = ast_strdupa(recdata->recording_ext);
```
Comment (jcolp) [inline apps/app_voicemail.c:6187]: This should be a separate PR since it's to fix IMAP building

File: apps/app_voicemail.c [hunk 19]
```diff
@@ -11413,6 +11412,365 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 6) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (remove_message_from_mailbox(a)){
+		ast_log(AST_LOG_WARNING, "Error removing message!\n");
```
Comment (jcolp) [inline apps/app_voicemail.c:11768]: Should use ast_cli, and it's nice to also include details and make it more readable - such as: "Error removing message X from mailbox Y of context Z"

**Review by jcolp — CHANGES_REQUESTED** [review]

File: apps/app_voicemail.c (hunk 18 above)
Comment (mbradeen) [reply apps/app_voicemail.c]: moved to it's own pr/issue

File: apps/app_voicemail.c (hunk 17 above)
Comment (mbradeen) [reply apps/app_voicemail.c]: Modified all three commands to output pass/fail with details.

File: apps/app_voicemail.c [hunk 20]
```diff
@@ -11413,6 +11413,389 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
 	return 0;
 }
 
+static int show_mailbox_details(struct ast_cli_args *a)
+{
+#define VMBOX_STRING_HEADER_FORMAT "%-32.32s %-32.32s %-16.16s %-16.16s %-16.16s %-16.16s\n"
+#define VMBOX_STRING_DATA_FORMAT   "%-32.32s %-32.32s %-16.16s %-16.16s %-16.16s %-16.16s\n"
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11419]: The values for the 2 macros are identical, do they need to be separate?

File: apps/app_voicemail.c [hunk 21]
```diff
@@ -11413,6 +11413,389 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+		ast_cli(a->fd, VM_STRING_HEADER_FORMAT, "Folder", "Caller ID", "Date", "Duration", "Flag", "ID");
+
+		for (int i = 0; i < mailbox_snapshot->folders; i++) {
+			AST_LIST_TRAVERSE(&((mailbox_snapshot)->snapshots[i]), msg, msg) {
+				ast_cli(a->fd, VM_STRING_DATA_FORMAT, msg->folder_name, msg->callerid, msg->origdate, msg->duration,
+						msg->flag, msg->msg_id);
+			}
+		}
+
+		ast_cli(a->fd, "%d Messages Total\n", mailbox_snapshot->total_msg_num);
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11459]: Use the ESS macro for messages to probably singularize/pluarlize it.

File: apps/app_voicemail.c [hunk 22]
```diff
@@ -11413,6 +11413,389 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+	show_mailbox_snapshot(a);
+	return 0;
+}
+
+static int forward_message_from_mailbox(struct ast_cli_args *a)
+{
+	const char *from_mailbox = a->argv[2];
+	const char *from_context = a->argv[3];
+	const char *from_folder = a->argv[4];
+	const char *id[] = {a->argv[5]};
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11482]: Add spaces inside the braces, it's easy to miss that that's initializing an array.

File: apps/app_voicemail.c [hunk 23]
```diff
@@ -11413,6 +11413,389 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+	const char *to_mailbox = a->argv[6];
+	const char *to_context = a->argv[7];
+	const char *to_folder = a->argv[8];
+	int ret = vm_msg_forward(from_mailbox, from_context, from_folder, to_mailbox, to_context, to_folder, 1, id, 0);
+	if (ret) {
+		ast_cli(a->fd, "Error forwarding message %s from mailbox %s@%s %s to mailbox %s@%s %s\n",
+					id[0], from_mailbox, from_context, from_folder, to_mailbox, to_context, to_folder);
+	}
+	else{
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11491]: needs to be on prev line

File: apps/app_voicemail.c [hunk 24]
```diff
@@ -11413,6 +11413,389 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+	const char *context = a->argv[3];
+	const char *from_folder = a->argv[4];
+	const char *id[] = {a->argv[5]};
+	const char *to_folder = a->argv[6];
+	int ret =  vm_msg_move(mailbox, context, 1, from_folder, id, to_folder);
+	if (ret) {
+		ast_cli(a->fd, "Error moving message %s from mailbox %s@%s %s to %s\n",
+					id[0], mailbox, context, from_folder, to_folder);
+	}
+	else{
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11510]: prev line

File: apps/app_voicemail.c [hunk 25]
```diff
@@ -11413,6 +11413,389 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+	const char *mailbox = a->argv[2];
+	const char *context = a->argv[3];
+	const char *folder = a->argv[4];
+	const char *id[] = {a->argv[5]};
+	int ret = vm_msg_remove(mailbox, context, 1, folder, id);
+	if (ret) {
+		ast_cli(a->fd, "Error removing message %s from mailbox %s@%s %s\n",
+					id[0], mailbox, context, folder);
+	}
+	else{
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11528]: prev line

File: apps/app_voicemail.c [hunk 26]
```diff
@@ -11413,6 +11413,389 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+		return complete_voicemail_show_mailbox(a);
+	case CLI_HANDLER:
+		break;
+	}
+
+	if (a->argc != 5) {
+		return CLI_SHOWUSAGE;
+	}
+
+	if (show_messages_for_mailbox(a)) {
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11601]: This function always returns 0, so this warning is unreachable. And as is, this warning message isn't really helpful since it lacks details.

File: apps/app_voicemail.c [hunk 27]
```diff
@@ -11413,6 +11413,389 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+					AST_LIST_UNLOCK(&users);
+					return ret;
+				}
+				mailbox = vmu->mailbox;
+			}
+		}
+		AST_LIST_UNLOCK(&users);
+	} else if (pos == 3 || pos == 7) {
+		/* find contexts that match the user */
+		if (pos == 3) {
```
Comment (InterLinked1) [inline apps/app_voicemail.c:11659]: You could tidy this up by using the ternary operator...

**Review by InterLinked1 — CHANGES_REQUESTED** [review]

File: apps/app_voicemail.c (hunk 26 above)
Comment (mbradeen) [reply apps/app_voicemail.c:11599]: changed the underlying behavior and to return the CLI error like the other commands.

File: apps/app_voicemail.c (hunk 20 above)
Comment (mbradeen) [reply apps/app_voicemail.c:11419]: This is to be consistent with the other methods in case the line format changes and in case we add a concise/verbose option.

File: apps/app_voicemail.c [hunk 28]
```diff
@@ -11413,6 +11413,382 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+{
+#define VM_STRING_HEADER_FORMAT "%-8.8s %-32.32s %-32.32s %-9.9s %-6.6s %-30.30s\n"
+#define VM_STRING_DATA_FORMAT   "%-8.8s %-32.32s %-32.32s %-9.9s %-6.6s %-30.30s\n"
+	const char *mailbox = a->argv[3];
+	const char *context = a->argv[4];
+	struct ast_vm_mailbox_snapshot *mailbox_snapshot;
+	struct ast_vm_msg_snapshot *msg;
+
+	/* Take a snapshot of the mailbox and walk through each folder's contents */
+	if ((mailbox_snapshot = ast_vm_mailbox_snapshot_create(mailbox, context, NULL, 0, AST_VM_SNAPSHOT_SORT_BY_ID, 0))) {
```
Comment (jcolp) [inline apps/app_voicemail.c:11449]: If there is no mailbox_snapshot then you can return early and reduce indentation

**Review by jcolp — CHANGES_REQUESTED** [review]

**Review by gtjoseph — APPROVED** [review]

File: apps/app_voicemail.c [hunk 29]
```diff
@@ -11418,6 +11418,383 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
...
+	/* Take a snapshot of the mailbox and walk through each folder's contents */
+	mailbox_snapshot = ast_vm_mailbox_snapshot_create(mailbox, context, NULL, 0, AST_VM_SNAPSHOT_SORT_BY_ID, 0);
+	if (!mailbox_snapshot) {
+		ast_cli(a->fd, "Can't create snapshot for voicemail user %s@%s\n", mailbox, context);
+		return -1;
+	}
+
+	ast_cli(a->fd, VM_STRING_HEADER_FORMAT, "Folder", "Caller ID", "Date", "Duration", "Flag", "ID");
+
+	for (int i = 0; i < mailbox_snapshot->folders; i++) {
```
Comment (gtjoseph) [inline apps/app_voicemail.c:11461]: oops, we missed these.  can't declare in for statement.
